contrib/codegen/src/database.rs

```diff
@@ -0,0 +1,110 @@
+    if !input.generics.params.is_empty() {
+        return Err(input.span().error(NO_GENERIC_STRUCTS));
+    }
+
+    let structure = match input.data {
+        Data::Struct(s) => s,
+        _ => return Err(input.span().error(ONLY_ON_STRUCTS_MSG))
+    };
+
+    let inner_type = match structure.fields {
+        Fields::Unnamed(ref fields) if fields.unnamed.len() == 1 => {
+            let first = fields.unnamed.first().expect("checked length");
+            first.value().ty.clone()
+        }
+        _ => return Err(structure.fields.span().error(ONLY_UNNAMED_FIELDS).help(EXAMPLE))
+    };
+
+    Ok(DatabaseInvocation {
+        type_name: input.ident,
+        db_name: string_lit.value(),
+        structure: structure,
+        connection_type: inner_type,
+    })
+}
+
+pub fn database_attr(attr: TokenStream, input: TokenStream) -> Result<TokenStream> {
+    let invocation = parse_invocation(attr, input)?;
+
+    let connection_type = &invocation.connection_type;
+    let database_name = &invocation.db_name;
+    let request_guard_type = &invocation.type_name;
+    let pool_type = Ident::new(&format!("{}Pool", request_guard_type), request_guard_type.span());
+
+    let tokens = quote! {
+        pub struct #request_guard_type(::rocket_contrib::databases::r2d2::PooledConnection<<#connection_type as ::rocket_contrib::databases::Poolable>::Manager>);
+        pub struct #pool_type(::rocket_contrib::databases::r2d2::Pool<<#connection_type as ::rocket_contrib::databases::Poolable>::Manager>);
+
+        impl #request_guard_type {
+            pub fn fairing() -> impl ::rocket::fairing::Fairing {
+                use ::rocket_contrib::databases::Poolable;
+
+                ::rocket::fairing::AdHoc::on_attach(|rocket| {
+                    let pool = ::rocket_contrib::databases::database_config(#database_name, &rocket.config()) // Option<DatabaseConfig<'a>>
+                        .map(#connection_type::pool); // Option<Result<_, _>>
+
+                    match pool {
+                        Some(Ok(e)) => Ok(rocket.manage(#pool_type(e))),
+                        Some(Err(_)) | None => Err(rocket),
+                    }
+                })
+            }
+        }
+
+        impl ::std::ops::Deref for #request_guard_type {
+            type Target = #connection_type;
+
+            #[inline(always)]
+            fn deref(&self) -> &Self::Target {
+                &self.0
+            }
+        }
+
+        impl<'a, 'r> ::rocket::request::FromRequest<'a, 'r> for #request_guard_type {
+            type Error = ();
+
+            fn from_request(request: &'a ::rocket::request::Request<'r>) -> ::rocket::request::Outcome<Self, Self::Error> {
+                let pool = request.guard::<::rocket::State<#pool_type>>()?;
+
+                match pool.0.get() {
+                    Ok(conn) => ::rocket::Outcome::Success(#request_guard_type(conn)),
+                    Err(_) => ::rocket::Outcome::Failure((Status::ServiceUnavailable, ())),
```

Comment:
Status should be fully qualified if everything is going to be fully qualified: `::rocket::http::Status`

Reply:
Yep, that one I missed.